### ISSUE TITLE
[31] Swin Transformer V2: Scaling Up Capacity and Resolution

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#59]
[paper](https://arxiv.org/pdf/2111.09883.pdf)
[code](https://github.com/microsoft/Swin-Transformer)

모델을 3B까지 키움.
# Abstract
![image](https://user-images.githubusercontent.com/16400591/142803143-3c85e961-e52e-4beb-ac03-b20fe543b2a7.png)

1. pre norm ==> post norm. (그렇지만 skip add 다음이 아니라 skip add 이전에 ln을 적용한 다음 add하는 방식.)
2. dot prod attn ==> scaled cos attn. attn logit 값을 normalize 하기 위함.
3. position을 linear에서 log space continuous position bias 로 변화
4. discrete learnable pos bias ==> 2 layer mlp를 사용한 continuous bias.

(1), (2) 는 모델 capacity 늘리는 데 도움이 됨. (3) 은 transferrability 에 영향을 줌 (resolution, window-size)

# INTRO
NLP 진영은 BERT (340M) 부터 Megatron (530 billion) 을 거쳐 Switch-Transformers (1.6 trillion) 까지, 엄청나게 모델이 커져가고 있음.
반대로, vision은 그대로임
vision 모델에서 큰 모델을 쓰려면 고려할 사항들이 많음
1. Instability
    1. layer 별 activation amplitude 가 큰 모델에서 더 차이가 심해짐
    2. 이유를 좀 더 살펴보니, output 에 그대로 더해주는 방식의 Skip-conn 때문이었음
    3. LN 위치를 뒤로 옮기는 post-norm 방식을 사용
    4. scaled cosine attention 을 써서 input 의 amplitude 에 영향이 없도록 만들고, attention value가 튀지 않도록 해줌
2. Transferability
    1. downstream 중에서도 Obeject-Detection 이나 Segmentation은 input resolution 도 크고 attention-window 도 커야 함
    2. Arbirtrary coordinate 들에 대해 bias value 를 생성하는 Log-CPB (log-spaced continuous position bias) 제안
3. Memory
    1. deepspeed 의 zero-optimizer 사용
    2. gradient checkpointing 사용
    3. sequential self-attention 계산 (말 그대로 sequential 하게 계산하는 건데, 첫 2개 layer 에만 적용함)

# Swin V2

## Recap swin v1
![image](https://user-images.githubusercontent.com/16400591/142825836-cfa311fa-1e28-4e2c-a270-fb301589d4a5.png)
유명한 swin v1. 

swin v1의 attention 은 다음 수식과 같이 수행함
![image](https://user-images.githubusercontent.com/16400591/142831219-11586fb6-6d55-4460-888a-ed06fb0f30b1.png)
B 는 learnable 한 녀석인데,
MxM patch 를 이용한다고 할 때,
relative 좌표들은 [-M+1, M-1] 범위의 값들을 x,y 에 대해서 각각 가진다.
그래서 (2M-1)x(2M-1) size 의 learnable parameter 를 만들어서 discrete 좌표들을 embedding 하는 역할로 사용한다.

더 이상 자세한 설명 생략.

다만, 몇 가지 짚을 점이 있음
![image](https://user-images.githubusercontent.com/16400591/142809388-fdad2305-15da-40ce-8b7f-6580fabf9c6e.png)
위 이미지는 Signal Propagation Plot (deepmind 논문 https://arxiv.org/pdf/2101.08692.pdf 참고) 을 해 본 것이다.
layer가 쌓일수록 굉장히 instable 하다.

H-Pre 는 * 이 붙어 있는데, 40 epoch 전에 학습이 망가졌다는 뜻이다.
학습 그래프는 다음과 같다.
![image](https://user-images.githubusercontent.com/16400591/142809368-0225ed77-7c29-4c17-965a-b1fe1ce889d0.png)

## Scaling Up Model Capacity
앞서 말했듯이, post-normalization 의 효과가 좋았다.
추가로, scaled cosine attention 을 제안했는데,
dot-product attention을 다음 식으로 대체했다.
![image](https://user-images.githubusercontent.com/16400591/142827320-0ef2ea7c-2966-4224-bcf0-1c5ae5a07ea1.png)
τ 는 learnable scalar 이고, layer나 head 별로 하나씩 존재한다 (share 안함)
B 는 relative positional bias 인데, swin v1 에서는 큰 resolution 으로 transfer를 수행할 때, 요 녀석을 scale 시켰었다.
아래 항목에서 swin v2 의 relative positional bias 사용 방식을 설명한다.

## Scaling Up Window Resolution
relative position ∆x, ∆y 는 우선 log-scale 로 바뀌어 진다.
![image](https://user-images.githubusercontent.com/16400591/142828564-f0fb7288-89e8-48ad-9b50-a62ea835b55b.png)
log-scale 로 바꿈으로써, window-size 가 바뀌며 transfer 를 할 때, 잘 동작하게 된다.
예를 들어, patch를 8x8 에서 16x16 으로 바꾸면, relative pos range 가 `[-7, 7]` 에서 `[-15, 15]` 로 바뀐다. extrapolation ratio 는 8개 를 채워야 하니, `8/7==1.14` 이 된다.
log scale range 를 사용한다면, `[−2.079, 2.079]` 에서 `[−2.773, 2.773]` 로 바뀌고, extrapolation ratio 는 `0.33`에 불과하다.

![image](https://user-images.githubusercontent.com/16400591/142828193-53feffdb-94de-4c9b-92c0-fe1610da8460.png)
구해낸 log-scale ∆x, ∆y 에 대해서 meta-network 를 태우면 끝난다.
meta network G 는 그냥 2-layer MLP 이다. (Light 하다)

이런 작업을 통해 다양한 window-size 를 해결해 낼 수 있다.
![image](https://user-images.githubusercontent.com/16400591/142809322-2cbbdd10-2c95-4d62-a213-ef44905e3265.png)

결과에서 보이듯이, Object-Detection이나 Segmentation에서의 성능향상이 두드러진다.

# Results
## Ablations
![image](https://user-images.githubusercontent.com/16400591/142809718-85c06344-cb9a-40ce-ae9f-0fabe469adb0.png)
![image](https://user-images.githubusercontent.com/16400591/142809733-32e58ca3-78bd-445c-97b3-3811280999f5.png)

## Classification
![image](https://user-images.githubusercontent.com/16400591/142809479-42c729d9-4517-4817-899e-de420b573e1c.png)
## Object Detection
![image](https://user-images.githubusercontent.com/16400591/142837978-075eff78-9ef8-48dc-a484-b37af5f8eae7.png)

## Semantic Segmentation
![image](https://user-images.githubusercontent.com/16400591/142809659-7be05c3e-d848-45a2-9a42-8f42d67952a6.png)

## Video Action Classification
![image](https://user-images.githubusercontent.com/16400591/142809703-faef8e6b-97c8-4425-a4e0-795744d47b38.png)
